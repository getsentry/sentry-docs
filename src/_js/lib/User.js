const ORIGIN = process.env.JEKYLL_GETSENTRY_ORIGIN || 'https://sentry.io';
import { escape } from './Helpers';
import { logPageview } from './Page';

export default class User {
  constructor() {
    this.namespace = 'docsUserDataV1';
    this.init = this.init.bind(this);
    this.update = this.update.bind(this);
    this.setUserData = this.setUserData.bind(this);
    this.getUserData = this.getUserData.bind(this);
  }

  update() {
    $(document).trigger('user.didUpdate', [this.getUserData()]);
  }

  setUserData(newData = {}) {
    let userData = {};

    try {
      userData = JSON.parse(localStorage.getItem(this.namespace)) || {};
    } catch (error) {
      // Don't worry about errors if localStorage is corrupt, we'll just discard
      // the object and start fresh.
    }

    const { projects: tmpProjects, preferred: tmpPref, user } = {
      ...userData,
      ...newData
    };

    const projects =
      tmpProjects && tmpProjects.length ? tmpProjects : [defaultProject];

    // Default to the first project
    let preferred = projects[0];
    if (tmpPref) {
      // If a preference is set, check it still exists, otherwise still use
      // the first project as the default.
      preferred = projects.find(({ id }) => id === tmpPref.id) || preferred;
    }

    const payload = { projects, preferred, user: user || defaultUser };
    localStorage.setItem(this.namespace, JSON.stringify(payload));
    return payload;
  }

  getUserData() {
    let data = null;
    try {
      data = JSON.parse(localStorage.getItem(this.namespace));
    } catch (error) {
      // If the parse failed, we'll just purge the data
    }

    const { preferred, projects, user } = data || this.setUserData();
    return {
      preferred: constructDSNObject(preferred),
      projects: projects.map(constructDSNObject),
      user
    };
  }

  setPreference(newID) {
    const { projects } = this.getUserData();
    const preferred = projects.find(({ id }) => id === newID);
    this.setUserData({ preferred });
    this.update();
  }

  init() {
    const cached = JSON.parse(localStorage.getItem(this.namespace));
    if (cached) this.update(cached);

    return $.ajax({
      type: 'GET',
      url: ORIGIN + '/docs/api/user/',
      crossDomain: true,
      xhrFields: {
        withCredentials: true
      }
    })
      .then(({ projects, user }) => {
        window.ra.identify(user.id);
        window.amplitude.getInstance().setUserId(user.id);
        const { isAuthenticated } = user;
        $('[data-hide-when-logged-in]').toggleClass('d-none', isAuthenticated);
        this.setUserData({ projects, user });
        this.update();
      })
      .catch(error => {
        $('[data-hide-when-logged-in]').toggleClass('d-none', false);
        this.update();
      })
      .then(() => {
        logPageview();
      });
  }
}

const formatDsn = function(
  { publicKey, secretKey, scheme, host, pathSection },
  opts = { public: false }
) {
  const auth = opts.public ? publicKey : `${publicKey}:${secretKey}`;
  return `${scheme}${auth}@${host}${pathSection}`;
};

const formatMinidumpURL = function(dsn) {
  const { scheme, host, pathSection, publicKey } = dsn;
  return `${scheme}${host}/api${pathSection}/minidump/?sentry_key=${publicKey}`;
};

const formatUnrealEngineURL = function(dsn) {
  const { scheme, host, pathSection, publicKey } = dsn;
  return `${scheme}${host}/api${pathSection}/unreal/${publicKey}/`;
};

const formatAPIURL = function(dsn) {
  return `${dsn.scheme}${dsn.host}/api`;
};

export const defaultUser = {
  isAuthenticated: false
};

export const defaultProject = {
  projectName: 'your-project',
  secretKey: '<secret>',
  publicKey: '<key>',
  dsnPublic: 'https://<key>@sentry.io/<project>',
  id: -1,
  dsn: 'https://<key>:<secret>@sentry.io/<project>',
  organizationSlug: 'your-org',
  projectSlug: 'your-project'
};

export const constructDSNObject = function(project = {}) {
  project = { ...defaultProject, ...project };
  // Transform the dsn into useful values
  const match = project.dsn.match(/^(.*?\/\/)(.*?):(.*?)@(.*?)(\/.*?)$/);

  const dsn = {
    scheme: escape(match[1]),
    publicKey: escape(match[2]),
    secretKey: `${escape(match[3])}`,
    host: escape(match[4]),
    pathSection: escape(match[5])
  };

  return {
    id: project.id,
    group: escape(project.organizationSlug),
    PROJECT_NAME: escape(project.projectSlug),
    PROJECT_ID: project.id.toString(),
    ORG_NAME: escape(project.organizationSlug),
    DSN: formatDsn(dsn),
    PUBLIC_DSN: formatDsn(dsn, { public: true }),
    PUBLIC_KEY: dsn.publicKey,
    SECRET_KEY: dsn.secretKey,
    API_URL: formatAPIURL(dsn),
    MINIDUMP_URL: formatMinidumpURL(dsn),
    UNREAL_URL: formatUnrealEngineURL(dsn)
  };
};
