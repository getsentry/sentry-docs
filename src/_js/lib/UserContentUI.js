// These are the variables that can be replaced in the document
export const tokens = [
  'DSN',
  'PUBLIC_DSN',
  'PUBLIC_KEY',
  'SECRET_KEY',
  'API_URL',
  'PROJECT_ID',
  'ORG_NAME',
  'PROJECT_NAME',
  'MINIDUMP_URL',
  'UNREAL_URL'
];

// Helper to generate a regex from the approved tokens
const tokenRegex = function() {
  var str = `___(${tokens.join('|')})___`;
  var re = new RegExp(str, 'g');
  return re;
};

const renderDropdownItem = function(item, selected) {
  return `<a href="#" class="dropdown-item${
    item.id === selected.id ? ' active' : ''
  }" data-id="${item.id}">${item.ORG_NAME} / ${item.PROJECT_NAME}</a>`;
};

const renderDropdownHeader = function(item) {
  return `<h6 class="dropdown-header">${item}</h6>`;
};

const renderDropdown = function(list, preferred) {
  const grouped = list.reduce((obj, item) => {
    if (obj[item.group] === undefined) obj[item.group] = [];
    obj[item.group].push(item);
    return obj;
  }, {});

  return `
    <div class="dropdown">
      <a class="dropdown-toggle" href="#" role="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false" data-not-dynamic>
        ${preferred.ORG_NAME} / ${preferred.PROJECT_NAME}
      </a>

      <div class="dropdown-menu user-content-menu" aria-hidden="true">
      ${Object.keys(grouped)
        .map(key => {
          return `${renderDropdownHeader(key)}${grouped[key]
            .map(item => renderDropdownItem(item, preferred))
            .join('')}`;
        })
        .join('')}
      </div>
    </div>
  `;
};

// Updates all tokens to show values based on the values stored in localStorage
const updateTokens = function({ projects, preferred }) {
  $('[data-user-content]').each(function(i, el) {
    const $el = $(el);
    const key = $el.data('user-content');
    $el.html(unescape(preferred[key]));
  });

  $('[data-user-content-switcher-mount]').each(function() {
    $(this).html(renderDropdown(projects, preferred));
  });

  $(document).trigger('userContent.didLoad');
};

const wrapTokens = function() {
  // Update UI with our widgets
  $('main pre').each(function(i, el) {
    const $el = $(el);

    if (!tokenRegex().test($el.html())) return true;

    // Wrap all tokens with an element that can be used for updating
    const content = $el.html().replace(tokenRegex(), function(match) {
      const token = match.replace(/___/g, '');
      return `<span data-user-content="${token}"></span>`;
    });
    $el.html(content);

    // Add a dropdown for switching the data.
    $el.before(`
      <div class="py-1 px-2 text-white" data-user-content-switcher>
        <small>Showing<span class="d-none d-sm-inline"> configuration for</span></small>
        <div data-user-content-switcher-mount></div>
      </div>
    `);
  });
};

const init = function() {
  $('[data-hide-when-logged-in]').addClass('d-none');

  // Update tokens whenever the User data updates
  $(document).on('user.didUpdate', function(event, userData) {
    wrapTokens();
    updateTokens(userData);

    $(document).on('page.didUpdate', function() {
      wrapTokens();
      updateTokens(userData);
    });
  });

  const ucSelector = '[data-user-content-switcher] .dropdown-item';
  $(document).on('click', ucSelector, function(event) {
    event.preventDefault();
    const id = $(event.target).data('id');
    User.setPreference(id);
  });
};

export default { init };
