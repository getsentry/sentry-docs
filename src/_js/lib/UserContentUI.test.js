import UserContentUI, { init } from './UserContentUI';
import User from './User';
import 'jest-localstorage-mock';

const MOCK_HTML = `
  <!DOCTYPE html>
  <html lang="en">

  <head>
    <title>This is a new document</title>
  </head>

  <body>
    <main>
      <div data-hide-when-logged-in>This should be hidden after page load</div>
      <pre><code>This is a basic example: ___PUBLIC_DSN___</code></pre>
      <pre><code>This example should not be wrapped</pre>
    </main>
  </body>

  </html>
`;

const sampleProject1 = {
  id: 1,
  group: 'test1',
  PROJECT_NAME: 'test1',
  PROJECT_ID: '1',
  ORG_NAME: 'test1',
  ORG_ID: '1',
  DSN: 'https://test1:test1@o1.ingest.sentry.io/test1',
  PUBLIC_DSN: 'https://test1@o1.ingest.sentry.io/test1',
  PUBLIC_KEY: 'test1',
  SECRET_KEY: 'test1',
  API_URL: 'https://sentry.io/api',
  MINIDUMP_URL:
    'https://o1.ingest.sentry.io/api/test1/minidump/?sentry_key=test1',
  MINIDUMP_URL: 'https://o1.ingest.sentry.io/api/test1/unreal/test1/'
};

const sampleProject2 = {
  id: 2,
  group: 'test1',
  PROJECT_NAME: 'test2',
  PROJECT_ID: '2',
  ORG_NAME: 'test2',
  ORG_ID: '2',
  DSN: 'https://test2:test2@o2.ingest.sentry.io/test2',
  PUBLIC_DSN: 'https://test2@o2.ingest.sentry.io/test2',
  PUBLIC_KEY: 'test2',
  SECRET_KEY: 'test2',
  API_URL: 'https://sentry.io/api',
  MINIDUMP_URL:
    'https://o2.ingest.sentry.io/api/test2/minidump/?sentry_key=test2',
  MINIDUMP_URL: 'https://o2.ingest.sentry.io/api/test2/unreal/test2/'
};

const sampleUser = {
  name: 'Test User',
  id: 1,
  isAuthenticated: true,
  avatarUrl: 'https://foo.bar.com'
};

describe('UserContentUI', function() {
  beforeEach(() => {
    const $ = require('jquery');
    global.$ = $;
    document.body.innerHTML = MOCK_HTML;
    global.User = new User();
    global.User.init();
  });

  describe('.init', () => {
    test('hide-when-logged-in', () => {
      expect($('[data-hide-when-logged-in]').attr('class')).toBe(undefined);
      UserContentUI.init();
      expect($('[data-hide-when-logged-in]').attr('class')).toBe('d-none');
    });

    test('user.didUpdate', () => {
      UserContentUI.init();
      $(document).trigger('user.didUpdate', [
        {
          preferred: sampleProject2,
          projects: [sampleProject1, sampleProject2]
        }
      ]);
      expect($('body').html()).toMatchSnapshot();
    });

    test('clicks on new projects switch the project', () => {
      UserContentUI.init();
      global.User.setPreference = jest.fn();
      $(document).trigger('user.didUpdate', [
        {
          preferred: sampleProject2,
          projects: [sampleProject1, sampleProject2]
        }
      ]);
      expect($('body').html()).toMatchSnapshot();
      $('.dropdown-item[data-id="1"]').click();
      expect(global.User.setPreference).toHaveBeenCalledWith(1);
    });

    test('rewraps after page update', () => {
      UserContentUI.init();
      global.User.setPreference = jest.fn();
      $(document).trigger('user.didUpdate', [
        {
          preferred: sampleProject2,
          projects: [sampleProject1, sampleProject2]
        }
      ]);
      document.body.innerHTML = MOCK_HTML;
      expect($('body').html()).toMatchSnapshot();
      $(document).trigger('page.didUpdate');
      expect($('body').html()).toMatchSnapshot();
    });
  });
});
