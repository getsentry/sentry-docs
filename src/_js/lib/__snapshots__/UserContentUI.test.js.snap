// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserContentUI .init clicks on new projects switch the project 1`] = `
"
  
  

  
    <title>This is a new document</title>
  

  
    <main>
      <div data-hide-when-logged-in=\\"\\" class=\\"d-none\\">This should be hidden after page load</div>
      
      <div class=\\"py-1 px-2 text-white\\" data-user-content-switcher=\\"\\">
        <small>Showing<span class=\\"d-none d-sm-inline\\"> configuration for</span></small>
        <div data-user-content-switcher-mount=\\"\\">
    <div class=\\"dropdown\\">
      <a class=\\"dropdown-toggle\\" href=\\"#\\" role=\\"button\\" data-toggle=\\"dropdown\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\" data-not-dynamic=\\"\\">
        test2 / test2
      </a>

      <div class=\\"dropdown-menu user-content-menu\\" aria-hidden=\\"true\\">
      <h6 class=\\"dropdown-header\\">test1</h6><a href=\\"#\\" class=\\"dropdown-item\\" data-id=\\"1\\">test1 / test1</a><a href=\\"#\\" class=\\"dropdown-item active\\" data-id=\\"2\\">test2 / test2</a>
      </div>
    </div>
  </div>
      </div>
    <pre><code>This is a basic example: <span data-user-content=\\"PUBLIC_DSN\\">https://test2@o2.ingest.sentry.io/test2</span></code></pre>
      <pre><code>This example should not be wrapped</code></pre><code>
    </code></main><code>
  

  
</code>"
`;

exports[`UserContentUI .init rewraps after page update 1`] = `
"
  
  

  
    <title>This is a new document</title>
  

  
    <main>
      <div data-hide-when-logged-in=\\"\\">This should be hidden after page load</div>
      <pre><code>This is a basic example: ___PUBLIC_DSN___</code></pre>
      <pre><code>This example should not be wrapped</code></pre><code>
    </code></main><code>
  

  
</code>"
`;

exports[`UserContentUI .init rewraps after page update 2`] = `
"
  
  

  
    <title>This is a new document</title>
  

  
    <main>
      <div data-hide-when-logged-in=\\"\\">This should be hidden after page load</div>
      
      <div class=\\"py-1 px-2 text-white\\" data-user-content-switcher=\\"\\">
        <small>Showing<span class=\\"d-none d-sm-inline\\"> configuration for</span></small>
        <div data-user-content-switcher-mount=\\"\\">
    <div class=\\"dropdown\\">
      <a class=\\"dropdown-toggle\\" href=\\"#\\" role=\\"button\\" data-toggle=\\"dropdown\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\" data-not-dynamic=\\"\\">
        test2 / test2
      </a>

      <div class=\\"dropdown-menu user-content-menu\\" aria-hidden=\\"true\\">
      <h6 class=\\"dropdown-header\\">test1</h6><a href=\\"#\\" class=\\"dropdown-item\\" data-id=\\"1\\">test1 / test1</a><a href=\\"#\\" class=\\"dropdown-item active\\" data-id=\\"2\\">test2 / test2</a>
      </div>
    </div>
  </div>
      </div>
    <pre><code>This is a basic example: <span data-user-content=\\"PUBLIC_DSN\\">https://test2@o2.ingest.sentry.io/test2</span></code></pre>
      <pre><code>This example should not be wrapped</code></pre><code>
    </code></main><code>
  

  
</code>"
`;

exports[`UserContentUI .init user.didUpdate 1`] = `
"
  
  

  
    <title>This is a new document</title>
  

  
    <main>
      <div data-hide-when-logged-in=\\"\\" class=\\"d-none\\">This should be hidden after page load</div>
      
      <div class=\\"py-1 px-2 text-white\\" data-user-content-switcher=\\"\\">
        <small>Showing<span class=\\"d-none d-sm-inline\\"> configuration for</span></small>
        <div data-user-content-switcher-mount=\\"\\">
    <div class=\\"dropdown\\">
      <a class=\\"dropdown-toggle\\" href=\\"#\\" role=\\"button\\" data-toggle=\\"dropdown\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\" data-not-dynamic=\\"\\">
        test2 / test2
      </a>

      <div class=\\"dropdown-menu user-content-menu\\" aria-hidden=\\"true\\">
      <h6 class=\\"dropdown-header\\">test1</h6><a href=\\"#\\" class=\\"dropdown-item\\" data-id=\\"1\\">test1 / test1</a><a href=\\"#\\" class=\\"dropdown-item active\\" data-id=\\"2\\">test2 / test2</a>
      </div>
    </div>
  </div>
      </div>
    <pre><code>This is a basic example: <span data-user-content=\\"PUBLIC_DSN\\">https://test2@o2.ingest.sentry.io/test2</span></code></pre>
      <pre><code>This example should not be wrapped</code></pre><code>
    </code></main><code>
  

  
</code>"
`;
